tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+	replicas tokenRingReplicas
+
+	// updating is true if we are currently updating the replica map for this keyspace.
+	updating bool
+
+	// needsUpdate is true if something changed while updating the replica map.
+	needsUpdate bool
+}
+
+// tokenAwareBanditHost keeps track of per-host state.
+type tokenAwareBanditHost struct {
+	// mu protects fields in tokenAwareBanditHost.
+	// If you need to hold both TokenAwareBanditHostPolicy.mu and tokenAwareBanditHost.mu,
+	// TokenAwareBanditHostPolicy.mu must be locked before tokenAwareBanditHost.mu.
+	mu sync.RWMutex
+	// latencyBuckets is a cyclic buffer storing history of latency measurements for a host.
+	// latencyBuckets[currentBucket] is the newest entry, the one to the right (with wrap) is the oldest entry.
+	latencyBuckets [bucketCount]latencyBucket
+	// currentBucket points to the newest bucket (the one we are writing to).
+	currentBucket int
+	// hostInfo of the host.
+	hostInfo *HostInfo
+}
+
+type latencyBucket struct {
+	sum   time.Duration
+	count int64
+}
+
+func (lb *latencyBucket) add(latency time.Duration) {
+	lb.sum += latency
+	lb.count += 1
+}
+
+// latency returns weighted average of latency measurements over time.
+// weight is linear from oldest bucket (1/bucketCount) to newest bucket (1.0).
+// h.mu must be held at least read-only.
+func (th *tokenAwareBanditHost) latency() (t time.Duration, ok bool) {
+	var sum time.Duration
+	var sumWeights int64
+	for i := 1; i <= len(th.latencyBuckets); i++ {
+		bucketIndex := (th.currentBucket + i) % bucketCount
+		if th.latencyBuckets[bucketIndex].count == 0 {
+			continue
+		}
+		sum += time.Duration(i) * th.latencyBuckets[bucketIndex].sum
+		sumWeights += int64(i) * th.latencyBuckets[bucketIndex].count
+	}
+	if sumWeights == 0 {
+		return 0, false
+	}
+	return sum / time.Duration(sumWeights), true
+}
+
+// recordLatency tracks the latency data point for this host.
+func (th *tokenAwareBanditHost) recordLatency(latency time.Duration) {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	th.latencyBuckets[th.currentBucket].add(latency)
+}
+
+func (th *tokenAwareBanditHost) decay() {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	th.currentBucket = (th.currentBucket + 1) % bucketCount
+	th.latencyBuckets[th.currentBucket] = latencyBucket{}
+}
+
+func (t *TokenAwareBanditHostPolicy) decay() {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	for _, host := range t.hosts {
+		host.decay()
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) decayLoop() {
+	ticker := time.NewTicker(t.decayPeriod / bucketCount)
+	for {
+		select {
+		case <-ticker.C:
+			t.decay()
+		case <-t.stopChan:
+			return
+		}
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) Init(s *Session) {
+	t.getKeyspaceMetadata = s.KeyspaceMetadata
+	go t.decayLoop()
+}
+
+func (t *TokenAwareBanditHostPolicy) Stop() {
+	t.stopOnce.Do(func() {
+		close(t.stopChan)
+	})
+}
+
+// hostInfoSlice converts t.hosts to a slice of *HostInfo.
+// It must be called with t.mu held for reading.
```

Comment:
Because we need to do multiple operations atomically. For example `resetRing` needs to replace token ring, and initiate update of replicas in all keyspaces. We can certainly move the locking into methods where we don't need to group multiple operations together.

Reply:
> We can certainly move the locking into methods where we don't need to group multiple operations together.

Let's do it where it's possible plz. I believe it will decrease the chance of introducing bugs after refactorings in the future.

Reply:
Already done in one of the previous commits, if you see any other place where this could be done, please let me know.